app/controllers/phases_controller.rb

```diff
@@ -1,44 +1,72 @@
+# frozen_string_literal: true
+
 class PhasesController < ApplicationController
-  before_action :set_phase, only: %i[show edit update destroy]
+  before_action :set_phase, only: %i[show edit update engineer destroy complete]
   before_action :set_project_lead
+  before_action :set_engineer, only: :show
 
   def index
     @phases = @project_lead.phases
   end
 
-  def show; end
+  def show
+    @users = User.joins(:roles).where('roles.name =? ', 2)
+  end
 
   def new
     @phase = @project_lead.phases.new
+    authorize @phase
   end
 
-  def edit; end
+  def edit
+    authorize @phase
+  end
 
   def create
     @phase = @project_lead.phases.new(phase_params)
+    authorize @phase
     return render :new, notice: 'Phase not Saved !' unless @phase.save
 
     redirect_to project_lead_phases_path, notice: 'Phase was successfully created.'
   end
 
+  def engineer
+    engineer = User.find(params[:engineer][:user_id])
+    if @phase.users.exists?(engineer.id)
+      flash[:alert] = 'Already Assigned'
+    else
+      @phase.users.append(engineer)
+      flash[:notice] = 'Engineer Assigned.'
+    end
+    redirect_to project_lead_phase_url(@phase.project_lead_id)
+  end
+
   def update
-    respond_to do |format|
-      if @phase.update(phase_params)
-        format.html { redirect_to project_lead_phases_url, notice: 'Phase Updated !'}
-        format.json { render :show, status: :ok, location: @phase }
-      else
-        format.html { render :edit, status: :unprocessable_entity }
-        format.json { render json: @phase.errors, status: :unprocessable_entity }
+    authorize @phase
+    if @phase.update(phase_params)
+      respond_to do |format|
+        format.html { redirect_to project_lead_phases_url, notice: 'Phase Updated !' }
       end
+    else render 'edit'
     end
   end
 
   def destroy
-    @phase.destroy
-    respond_to do |format|
-      format.html { redirect_to project_lead_phases_url, notice: 'Phase Deleted !' }
-      format.json { head :no_content }
+    authorize @phase
+    if @phase.destroy
+      respond_to do |format|
+        format.html { redirect_to project_lead_phases_url, notice: 'Phase Deleted !' }
+      end
+    end
+  end
+
+  def complete
+    if @phase.inactive!
+      flash[:notice] = 'Phase marked as complete successfully'
+    else
+      flash[:alert] = 'Phase can not be mark as complete '
     end
```

Comment:
This validation should be in model

Reply:
Its an action here because in Phase show i have a mark as complete link which is using this action